src/baldr/incident_singleton.h

```diff
@@ -46,23 +48,32 @@ struct incident_singleton_t {
   // daemon thread to watch for new incident data
   std::thread watcher;
 
+  // prototype for the watch function. we need this so unit tests can safely test all functionality
+  using watch_function_t = std::function<void(boost::property_tree::ptree,
+                                              std::unordered_set<valhalla::baldr::GraphId>,
+                                              std::shared_ptr<state_t>,
+                                              std::function<bool(size_t)>)>;
+
   /**
    * Singleton private constructor that static function uses to instantiate the singleton
-   * @param config   lets the daemon thread know where/how to look for incidents
-   * @param tileset  an mmapped graph tileset (ie static) allows incident loading to be lock-free
+   * @param config      lets the daemon thread know where/how to look for incidents
+   * @param tileset     an mmapped graph tileset (ie static) allows incident loading to be lock-free
+   * @param watch_func  the function the background thread will run to keep incident caches up to date
    */
   incident_singleton_t(const boost::property_tree::ptree& config,
-                       const std::unordered_set<valhalla::baldr::GraphId>& tileset)
-      : state{new state_t{}}, watcher(watch, config, tileset, state, interrupt()) {
+                       const std::unordered_set<valhalla::baldr::GraphId>& tileset,
+                       const watch_function_t& watch_func = incident_singleton_t::watch)
+      : state{new state_t{}}, watcher(watch_func, config, tileset, state, interrupt()) {
     // let the thread control its own lifetime
     watcher.detach();
     // check how long we should wait to find out if its initialized
     auto max_loading_latency =
         config.get<time_t>("incident_max_loading_latency", DEFAULT_MAX_LOADING_LATENCY);
+
```

Comment:
if you added a bit of code here like:

```c++
std::this_thread::sleep_for(std::chrono::seconds(1));
```

This would cause the watcher thread in the background to finish initializing and to notify BEFORE we ever got to the `cv::wait` below. It turns out that does doesnt work :smile:  if the condition variable is notified before you wait on it you'll never hear about it. basically the guidance is, never wait on just the condition variable always add a flag and a predicate that wraps it. in this way the predicate will always be checked before waiting (and bail early if already true) and after the time out or notify occurs returning whatever value the predicate returns.

Now the only way we can hit that throw is if the incident watching thread actually takes longer than the configured timeout to say that its initialized, which should be never (unless you have billions of incident tiles and a really tight timeout, which the unti tests dont have).